shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -154,34 +202,71 @@ private void validateAppend(Append append, Session session) {
         if (append.isConditional()) {
             throw new IllegalArgumentException("Conditional appends should be written via a ConditionalAppend object.");
         }
-        Preconditions.checkState(bytesLeftInBlock == 0 || bytesLeftInBlock > TYPE_PLUS_LENGTH_SIZE,
-                "Bug in CommandEncoder.encode, block is too small.");
     }
 
-    private void breakFromAppend(ByteBuf data, ByteBuf pendingData, ByteBuf out) {
-        if (bytesLeftInBlock != 0) {
+    private boolean isChannelFree() {
+        return writerIdPerformingAppends == null;
+    }
+
+    private boolean isChannelOwner(UUID writerID, String segment) {
+        return writerID.equals(writerIdPerformingAppends) && segment.equals(segmentBeingAppendedTo);
+    }
+
+    private boolean startAppend(ChannelHandlerContext ctx, AppendBatchSizeTracker blockSizeSupplier, Append append, ByteBuf out) {
+        final int msgSize = append.getData().readableBytes();
+        currentBlockSize = msgSize;
+        if (blockSizeSupplier != null) {
+            currentBlockSize = Math.max(currentBlockSize, blockSizeSupplier.getAppendBlockSize());
+        }
+        bytesLeftInBlock = currentBlockSize;
+        segmentBeingAppendedTo = append.segment;
+        writerIdPerformingAppends = append.writerId;
+        if (ctx != null && currentBlockSize > msgSize) {
+            writeMessage(new AppendBlock(writerIdPerformingAppends), currentBlockSize + TYPE_PLUS_LENGTH_SIZE, out);
+            ctx.executor().schedule(new BlockTimeouter(ctx.channel(), tokenCounter.incrementAndGet()),
+                    blockSizeSupplier.getBatchTimeout(),
+                    TimeUnit.MILLISECONDS);
+            return true;
+        }
+        return false;
+    }
+
+    private void continueAppend(ByteBuf data, ByteBuf out) {
+        bytesLeftInBlock -= data.readableBytes();
+        out.writeBytes(data);
+    }
+
+    private void breakFromAppend(ByteBuf data, ByteBuf pendingData,  boolean isAppend, ByteBuf out) {
+        if (isChannelFree()) {
+            return;
+        }
+        if (isAppend) {
```

Comment:
This boolean is always a constant passed at the call site. 
It would be better to have two methods instead. They can both share the code below the if by making that into a separate method.

Reply:
yes, its refactored now